assets/src/edit-story/components/library/panes/media/mediaPane.js

```diff
@@ -182,69 +222,97 @@ function MediaPane(props) {
-          </Title>
-          <MainButton onClick={openMediaPicker}>
-            {__('Upload', 'web-stories')}
-          </MainButton>
-        </Header>
-
-        <SearchInput
-          value={searchTerm}
-          placeholder={__('Search media...', 'web-stories')}
-          onChange={onSearch}
-        />
-
-        <FilterButtons>
-          {FILTERS.map(({ filter, name }, index) => (
-            <FilterButton
-              key={index}
-              active={filter === mediaType}
-              onClick={onFilter(filter)}
-            >
-              {name}
-            </FilterButton>
-          ))}
-        </FilterButtons>
-
-        {isMediaLoaded && !media.length ? (
-          <Message>{__('No media found', 'web-stories')}</Message>
-        ) : (
-          <Container>
-            <Column>
-              {resources
-                .filter((_, index) => isEven(index))
-                .map((resource) => (
-                  <MediaElement
-                    resource={resource}
-                    key={resource.src}
-                    width={DEFAULT_WIDTH}
-                    onInsert={insertMediaElement}
-                  />
-                ))}
-            </Column>
-            <Column>
-              {resources
-                .filter((_, index) => !isEven(index))
-                .map((resource) => (
-                  <MediaElement
-                    resource={resource}
-                    key={resource.src}
-                    width={DEFAULT_WIDTH}
-                    onInsert={insertMediaElement}
-                  />
-                ))}
-            </Column>
-          </Container>
-        )}
+        <Inner>
+          <Padding>
+            <Header>
+              <Title>
+                {__('Media', 'web-stories')}
+                {(!isMediaLoaded || isMediaLoading) && <Spinner />}
+              </Title>
+              <MainButton onClick={openMediaPicker}>
+                {__('Upload', 'web-stories')}
+              </MainButton>
+            </Header>
+
+            <SearchInput
+              value={searchTerm}
+              placeholder={__('Search media...', 'web-stories')}
+              onChange={onSearch}
+            />
+
+            <FilterButtons>
+              {FILTERS.map(({ filter, name }, index) => (
+                <FilterButton
+                  key={index}
+                  active={filter === mediaType}
+                  onClick={onFilter(filter)}
+                >
+                  {name}
+                </FilterButton>
+              ))}
+            </FilterButtons>
+          </Padding>
+
+          {isMediaLoaded && !media.length ? (
+            <Message>{__('No media found', 'web-stories')}</Message>
+          ) : (
+            <Container ref={refContainer}>
+              <Column>
+                {resources
+                  .filter((_, index) => isEven(index))
+                  .map((resource, i) => (
+                    <MediaElement
```

Comment:
My first reaction to this is that, most like `loading=lazy` will no longer suffice here. Both because it's not well supported, and also because we'd likely want to add more logic to visible vs not visible state. It's definitely fine to do this over several pull requests, of course. To highlight here: much of the performance tradeoffs will come from the `<video>` where we have more granularity: we can load the video, or just metadata, etc.

And looking at the current `MediaElement`, this would also improve the structure of that component(s). For instance, we can split it into two tiers: `MediaElementWrapper` that takes care of the sizing, focus, hover, etc. And a type-specific `Preview` component that's only rendered when the element is "visible" and can use system resources as it sees fit. This would also make a good place to deal with "poster backfill" and any such issue.